implementation/src/main/java/io/smallrye/jwt/auth/AbstractDpopTokenValidator.java

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright 2020 Red Hat, Inc, and individual contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.smallrye.jwt.auth;
+
+import static org.jose4j.jwa.AlgorithmConstraints.ConstraintType.PERMIT;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Map;
+
+import org.eclipse.microprofile.jwt.Claims;
+import org.eclipse.microprofile.jwt.JsonWebToken;
+import org.jose4j.jwa.AlgorithmConstraints;
+import org.jose4j.jwt.JwtClaims;
+import org.jose4j.jwt.consumer.InvalidJwtException;
+import org.jose4j.jwt.consumer.JwtConsumer;
+import org.jose4j.jwt.consumer.JwtConsumerBuilder;
+import org.jose4j.jwt.consumer.Validator;
+import org.jose4j.jwx.JsonWebStructure;
+import org.jose4j.keys.resolvers.EmbeddedJwkVerificationKeyResolver;
+import org.jose4j.lang.HashUtil;
+import org.jose4j.lang.JoseException;
+
+import io.smallrye.jwt.auth.principal.JWTAuthContextInfo;
+import io.smallrye.jwt.auth.principal.ParseException;
+
+/**
+ * Common functionality for classes implementing DPoP token proofing using
+ * HTTP request headers and a valid DPoP-bound {@link JsonWebToken}.
+ *
+ * @author Aaron Coburn {@literal <acoburn@apache.org>}
+ */
+public abstract class AbstractDpopTokenValidator {
+
+    protected static final String DPOP_HEADER = "DPoP";
+    protected static final String DPOP_HTTP_URI_CLAIM = "htu";
+    protected static final String DPOP_HTTP_METHOD_CLAIM = "htm";
+    protected static final String DPOP_JWT_TYPE = "dpop+jwt";
+    protected static final String JSON_WEB_KEY_THUMBPRINT = "jkt";
+
+    private final JWTAuthContextInfo authContextInfo;
+
+    protected AbstractDpopTokenValidator(JWTAuthContextInfo authContextInfo) {
+        this.authContextInfo = authContextInfo;
+    }
+
+    public void verify(JsonWebToken accessToken) throws ParseException {
```

Comment:
The idea here is that the verified access token is injected as `JsonWebToken` in a filter (JAX-RS, etc), correct ?

Reply:
The idea is that the access token needs to be present before the DPoP verification happens. An alternative would be to rely on an injected `JsonWebToken` in this class.

Reply:
Interesting, we can finalize on it a bit later, ideally one would just let `smallrye-jwt` to get `JsonWebToken` created instead of the code extending this abstract class creating it itself somehow. But thinking about it, your approach is more flexible, the extension can be a application or request scoped bean and have it injected anyway and then passed to this method...